### PR TITLE
Replace Goutte by Guzzle and adding testing suite

### PR DESCRIPTION
**Pull request by peter279k**
## Change log

1. replace the ``goutte`` with the ```Guzzle```
2. finish the unit testing. The testing result and code-coverage is available for the following url:
    https://travis-ci.org/peter279k/YahooWeatherAPI

    https://codecov.io/gh/peter279k/YahooWeatherAPI

### CODE REVIEW

**Comment by Th3Mouk** [comment]
Thanks for this amazing PR @peter279k, i will review it and make some annotations to merge it as soon as possible.

File: composer.json [hunk 1]
```diff
@@ -13,14 +13,18 @@
...
         }
     ],
     "require": {
         "php": "^5.6 || ^7.0",
-        "fabpot/goutte": "^3.1"
+        "guzzleHttp/guzzle": "^6.2"
     },
     "require-dev": {
-        "phpunit/phpunit": "^5.4.3",
+        "phpunit/phpunit": "^5.5",
```
Comment (Th3Mouk) [inline composer.json]: Why require this minor upgrade of version ?

File: phpunit.xml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Comment (Th3Mouk) [inline phpunit.xml]: Could you please format the xml to 4 spaces indent please ?

File: .travis.yml [hunk 3]
```diff
@@ -3,41 +3,17 @@ language: php
...
-  - mkdir -p ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d && echo "memory_limit=-1" >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
-  - mkdir -p build/logs
-  - composer global require satooshi/php-coveralls:@stable --no-update
-  - composer global update --prefer-dist --no-interaction
-  - composer update --prefer-dist --no-interaction $COMPOSER_FLAGS
+  - composer install
 
-before_script:
-  - echo "zend_extension=xdebug.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
+notifications:
```
Comment (Th3Mouk) [inline .travis.yml]: This got to be remove to be merge, why removing coveralls ?

File: .travis.yml [hunk 4]
```diff
@@ -3,41 +3,17 @@ language: php
 php:
   - 5.6
   - 7.0
-  - nightly
```
Comment (Th3Mouk) [inline .travis.yml]: Don't remove nightly from test, because we need to be proactive on this

File: .travis.yml [hunk 5]
```diff
@@ -3,41 +3,17 @@ language: php
...
-env:
-  global:
-    - PATH="$HOME/.composer/vendor/bin:$PATH"
-
-cache:
-  directories:
-    - $HOME/.composer/cache
-
-matrix:
-  fast_finish: true
```
Comment (Th3Mouk) [inline .travis.yml]: Fast finish must be preserved

File: composer.json (hunk 1 above)
Comment (peter279k) [inline composer.json]: Accodring to the [PHPUnit](https://phpunit.de/) official website,the version 5.5 is now currently released and it supports the PHP 7 and PHP5.6.

File: .travis.yml (hunk 4 above)
Comment (peter279k) [inline .travis.yml]: ok, I will fix it.

File: .travis.yml (hunk 5 above)
Comment (peter279k) [inline .travis.yml]: ok, I will fix it.

File: composer.json (hunk 1 above)
Comment (Th3Mouk) [inline composer.json]: If in none of the test needs a feature in this [changelog](https://github.com/sebastianbergmann/phpunit/blob/5.5/ChangeLog-5.5.md) please keep 5.4.3, we must remain permissive whenever possible.
The support of PHP5.6 and 7 is bring by 5.0.

File: .travis.yml (hunk 3 above)
Comment (peter279k) [inline .travis.yml]: ok, I will fix it. I remove too quick to remove some important code.

File: composer.json (hunk 1 above)
Comment (peter279k) [inline composer.json]: ok, why the PHPUnit version is ^5.4.3 ?
Can I change the PHPUnit version to ^5.4 ?

File: .travis.yml [hunk 6]
```diff
@@ -32,12 +32,18 @@ install:
   - composer global require satooshi/php-coveralls:@stable --no-update
   - composer global update --prefer-dist --no-interaction
   - composer update --prefer-dist --no-interaction $COMPOSER_FLAGS
+  - composer install
```
Comment (Th3Mouk) [inline .travis.yml]: This line isn't needed, the composer update does the same thing above

File: .travis.yml [hunk 7]
```diff
@@ -32,12 +32,18 @@ install:
...
+
+notifications:
```
Comment (Th3Mouk) [inline .travis.yml]: This need to be removed for the merge

File: phpunit.xml [hunk 8]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Th3Mouk) [inline phpunit.xml]: Please make an 4 space indent of xml files

File: tests/YahooWeatherClientTests.php [hunk 9]
```diff
@@ -10,11 +10,314 @@
 
 namespace Th3Mouk\YahooWeatherAPI\Tests;
 
+require 'src/Query/Query.php';
```
Comment (Th3Mouk) [inline tests/YahooWeatherClientTests.php]: Why do u use `require` instead of `use` ?

File: .travis.yml [hunk 10]
```diff
@@ -36,8 +36,6 @@ install:
 before_script:
   - echo "zend_extension=xdebug.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
 
-script:
-  - vendor/bin/phpunit -c phpunit.xml.dist --coverage-clover build/logs/clover.xml
+script: ./vendor/bin/phpunit tests/YahooWeatherClientTests.php --coverage-html --coverage-clover=coverage.xml
```
Comment (Th3Mouk) [inline .travis.yml]: this line can be reduced to 
```
- vendor/bin/phpunit -c phpunit.xml
```
Because all the configuration is now in the XML, but i don't know if `-c` option is always necessary

File: .gitignore [hunk 11]
```diff
@@ -1,3 +1,7 @@
 vendor
 .php_cs.cache
 composer.lock
+composer.phar
+clover.xml
+report
+result.txt
```
Comment (Th3Mouk) [inline .gitignore]: Remove this line and add a EOF please

File: phpunit.xml [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    <filter>
+	    <whitelist processUncoveredFilesFromWhitelist="true">
+		    <directory suffix=".php">./src/Query/Query.php</directory>
+            <directory suffix=".php">./src/YahooWeatherAPI.php</directory>
+            <directory suffix=".php">./src/YahooWeatherAPIInterface.php</directory>
+	    </whitelist>
+    </filter>
+	
+    <logging>
+	    <log type="coverage-clover" target="clover.xml"/>
```
Comment (Th3Mouk) [inline phpunit.xml]: indentation mistake

File: phpunit.xml [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+        <log type="coverage-text" target="result.txt" showUncoveredFiles="true"/>
```
Comment (Th3Mouk) [inline phpunit.xml]: This line isn't necessary, the result is prompted in CLI

File: phpunit.xml [hunk 14]
```diff
@@ -19,7 +19,7 @@
     </filter>
 	
     <logging>
-	    <log type="coverage-clover" target="clover.xml"/>
+	<log type="coverage-clover" target="clover.xml"/>
```
Comment (Th3Mouk) [inline phpunit.xml]: still here 😄

File: tests/YahooWeatherClientTests.php (hunk 9 above)
Comment (peter279k) [inline tests/YahooWeatherClientTests.php]: I just want to quick testing so I use require to load the all necessary PHP files.

File: .travis.yml [hunk 15]
```diff
@@ -37,7 +37,7 @@ before_script:
   - echo "zend_extension=xdebug.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
 
 script:
-  - vendor/bin/phpunit -c phpunit.xml.dist --coverage-clover build/logs/clover.xml
+  - vendor/bin/phpunit -c phpunit.xml
 
-after_script:
-  - coveralls -v
+after_success:
```
Comment (Th3Mouk) [inline .travis.yml]: Revert the changes please

File: tests/YahooWeatherClientTests.php (hunk 9 above)
Comment (Th3Mouk) [inline tests/YahooWeatherClientTests.php]: Ok feel free to correct the class in the same time

File: phpunit.xml [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    colors="true"
+    bootstrap="./vendor/autoload.php" type="coverage-text" target="tests/" charset="UTF-8">
+    
+    <testsuites>
+        <testsuite name="testing">
+      	    <directory>tests/</directory>
+        </testsuite>
+    </testsuites>
+
+    <filter>
```
Comment (Th3Mouk) [inline phpunit.xml]: Please have a look on this bundle for a filter example : https://github.com/Th3Mouk/YahooWeatherBundle/blob/master/phpunit.xml.dist#L21

Is this can be `<directory>./src</directory>` ?

IMO the previous code was good, just remove the useless exclusion rules.

File: phpunit.xml [hunk 17]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<phpunit
+    colors="true"
+    bootstrap="./vendor/autoload.php" type="coverage-text" target="tests/" charset="UTF-8">
+    
+    <testsuites>
```
Comment (Th3Mouk) [inline phpunit.xml]: The previous test suite don't need to be modified IMO

File: .travis.yml (hunk 15 above)
Comment (peter279k) [inline .travis.yml]: ok, I will add this setting back.

File: .travis.yml (hunk 10 above)
Comment (peter279k) [inline .travis.yml]: It;s very strange whem I use ```vendor/bin/phpunit -c phpunit.xml```.
The PHPUnit will not execute the unit testing and output the message: ```No tests executed``` during the travis-ci runtime.
And I got the same result from my local unit testing environment.

File: phpunit.xml (hunk 16 above)
Comment (peter279k) [inline phpunit.xml]: Ok.I modify some exclude tag in the new commit.

File: composer.json (hunk 1 above)
Comment (Th3Mouk) [inline composer.json]: Nop because in the 5.4.3 there's an important fix in the library concerning namespaces

File: .travis.yml [hunk 18]
```diff
@@ -37,7 +37,10 @@ before_script:
   - echo "zend_extension=xdebug.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
 
 script:
-  - vendor/bin/phpunit -c phpunit.xml.dist --coverage-clover build/logs/clover.xml
+  - vendor/bin/phpunit -c phpunit.xml
```
Comment (Th3Mouk) [inline .travis.yml]: I've just check this script configuration can be totally removed

See exemple 5.2 https://phpunit.de/manual/current/en/organizing-tests.html

File: .travis.yml [hunk 19]
```diff
@@ -37,7 +37,10 @@ before_script:
...
 
 after_script:
   - coveralls -v
+
+after_success:
```
Comment (Th3Mouk) [inline .travis.yml]: this config isn't needed anymore

File: .travis.yml (hunk 18 above)
Comment (peter279k) [inline .travis.yml]: Ok, the PHPUnit does not need ```-c phpunit.xml``` and it can automatically detect the phpunit.xml file.
I just use the command: ```vendor/bin/phpunit```

File: .travis.yml (hunk 19 above)
Comment (peter279k) [inline .travis.yml]: Did  you mean this config: after_success ? Don't you need the code coverage ?

File: .travis.yml (hunk 10 above)
Comment (Th3Mouk) [inline .travis.yml]: @peter279k U need to rename the class and the file to YahooWeatherClientApiTest.php

File: tests/YahooWeatherClientTests.php (hunk 9 above)
Comment (peter279k) [inline tests/YahooWeatherClientTests.php]: This problem has been solved when I modify the correct phpunit.xml file finally.

File: .travis.yml (hunk 10 above)
Comment (peter279k) [inline .travis.yml]: Ok, the PHPUnit does not need ```-c phpunit.xml``` and it can automatically detect the phpunit.xml file.
I just use the command: ```vendor/bin/phpunit```

Comment (Th3Mouk) [inline .travis.yml]: If U mainly PHPUnit i recommand an composer global install of the library. So you will only type `phpunit` 😉

File: .travis.yml (hunk 19 above)
Comment (Th3Mouk) [inline .travis.yml]: The code coverage is executed by PHPUnit. The after_script and `coveralls` command is usefull to send data to the inspector. Don't need to add an after_success command to travis

File: .gitignore (hunk 11 above)
Comment (Th3Mouk) [inline .gitignore]: Can you remove the 3 last lines please add add after the vendor, build

File: phpunit.xml (hunk 17 above)
Comment (Th3Mouk) [inline phpunit.xml]: Are the target and type properties relevant here ? Please keep the line break format foreach properties

File: phpunit.xml (hunk 12 above)
Comment (Th3Mouk) [inline phpunit.xml]: Replace by `build/logs/clover.xml`. All builds must be made in this folder

File: phpunit.xml [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+		    <directory suffix=".php">./src/Query/Query.php</directory>
+            <directory suffix=".php">./src/YahooWeatherAPI.php</directory>
+            <directory suffix=".php">./src/YahooWeatherAPIInterface.php</directory>
+	    </whitelist>
+    </filter>
+	
+    <logging>
+	    <log type="coverage-clover" target="clover.xml"/>
+        <log type="coverage-text" target="result.txt" showUncoveredFiles="true"/>
+        <log type="coverage-html" target="report" lowUpperBound="35" highLowerBound="70"/>
```
Comment (Th3Mouk) [inline phpunit.xml]: This line isn't necessary to, you get the result into the github hook

**Comment by Th3Mouk** [comment]
I quit internet for 72 hours in few hours, don't be worried if i don't reply.
IMO you got all the keys to be merged into monday.

File: .gitignore (hunk 11 above)
Comment (peter279k) [inline .gitignore]: Ok, I check this [link](https://github.com/Th3Mouk/YahooWeatherAPI/blob/master/.gitignore) to remove some additional lines.

File: phpunit.xml (hunk 20 above)
Comment (peter279k) [inline phpunit.xml]: Ok, I remove it and only this line: ```<log type="coverage-clover" target="clover.xml"/>``` here.

File: phpunit.xml (hunk 12 above)
Comment (peter279k) [inline phpunit.xml]: Ok, I will fix it.

File: composer.json [hunk 21]
```diff
@@ -13,11 +13,15 @@
...
+        },
+        {
+            "name": "Peter Lee",
+            "email": "peter279k@gmail.com"
         }
     ],
     "require": {
         "php": "^5.6 || ^7.0",
-        "fabpot/goutte": "^3.1"
+        "guzzleHttp/guzzle": "^6.2"
```
Comment (Th3Mouk) [inline composer.json]: please remove uppercase

File: .travis.yml (hunk 18 above)
Comment (Th3Mouk) [inline .travis.yml]: Try to remove the parameter

File: phpunit.xml [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+    
+    <testsuites>
+        <testsuite name="testing">
+      	    <directory suffix=".php">tests/</directory>
+        </testsuite>
+    </testsuites>
+
+    <filter>
+	    <whitelist>
+		    <directory suffix=".php">src/</directory>
```
Comment (Th3Mouk) [inline phpunit.xml]: Fix the indent please, you can param your IDE for future projects

**Comment by peter279k** [comment]
After I restart the build and It's done. I think It's the Network connection problem and it get the Imperfect response json.
[Travis-CI](https://travis-ci.org/peter279k/YahooWeatherAPI/builds/162412307)

File: phpunit.xml [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+    charset="UTF-8">
+    
+    <testsuites>
+        <testsuite name="testing">
+      	    <directory suffix=".php">tests/</directory>
+        </testsuite>
+    </testsuites>
+
+    <filter>
+	    <whitelist>
```
Comment (Th3Mouk) [inline phpunit.xml]: Finally this node indent and seems good for the phpunit

**Comment by Th3Mouk** [comment]
The error in the build seems arise from the PHPUnit version. With the lowest dependencies, he use 0.9 while he must use 5.4.3.

**Comment by Th3Mouk** [comment]
Could u run please a `php-cs-fixer fix .` command, the project contains some errors, I duno why they aren't  throwned by Style CI. I will have a look asap.

**Comment by Th3Mouk** [comment]
The fixer is fixed 😄 , could you rebase please

File: phpunit.xml (hunk 8 above)
Comment (peter279k) [inline phpunit.xml]: Hi @Th3Mouk , did you use [EditorConfig](http://editorconfig.org/#example-file) to solve  the indent prpblem?

**Comment by peter279k** [comment]
Hi @Th3Mouk , when I run the command: ```php-cs-fixer fix``` I get some error message: ``` Call to undefined method Symfony\CS\Config\Config::setRules() ``` . How should I fix this problem ? Thanks.

File: phpunit.xml (hunk 8 above)
Comment (Th3Mouk) [inline phpunit.xml]: Not yet, but i will have a look this week

**Comment by Th3Mouk** [comment]
There's still 72 commits

**Comment by peter279k** [comment]
Hi , I squash all commits now.

**Comment by Th3Mouk** [comment]
Could you rebase please, there's something wrong

**Comment by Th3Mouk** [comment]
Rebase into this [commit](https://github.com/Th3Mouk/YahooWeatherAPI/commit/de770023ff76f85cce9b19a7077e076211adca58) and after I will be able to merge

**Comment by Th3Mouk** [comment]
And probably this will reopen the PR

**Comment by peter279k** [comment]
How could I rebase this [commit](https://github.com/Th3Mouk/YahooWeatherAPI/commit/de770023ff76f85cce9b19a7077e076211adca58) ? I use the ```git rebase --root -i``` to squash my all commits.

**Comment by Th3Mouk** [comment]
See #4
